flow/probes/ebpf.go

```diff
@@ -80,298 +71,24 @@ type EBPFProbesHandler struct {
-				path, app := flow.LayersPath(p.Layers())
-				f.LayersPath += "/" + path
-				f.Application = app
-			}
-			*/
-		}
-	}
-
-	// ICMP
-	if layersInfo&uint8(C.ICMP_LAYER_INFO) > 0 {
-		kind := uint8(kernFlow.icmp_layer.kind)
-		code := uint8(kernFlow.icmp_layer.code)
-		id := uint16(kernFlow.icmp_layer.id)
-
-		f.ICMP = &flow.ICMPLayer{
-			Code: uint32(code),
-			ID:   uint32(id),
-		}
-		if f.Network != nil && f.Network.Protocol == flow.FlowProtocol_IPV4 {
-			f.ICMP.Type = flow.ICMPv4TypeToFlowICMPType(kind)
-		} else {
-			f.ICMP.Type = flow.ICMPv6TypeToFlowICMPType(kind)
-		}
-	}
-
-	appLayers := strings.Split(f.LayersPath, "/")
-	f.Application = appLayers[len(appLayers)-1]
-
-	f.Metric = &flow.FlowMetric{
-		ABBytes:   int64(kernFlow.metrics.ab_bytes),
-		ABPackets: int64(kernFlow.metrics.ab_packets),
-		BABytes:   int64(kernFlow.metrics.ba_bytes),
-		BAPackets: int64(kernFlow.metrics.ba_packets),
-		Start:     f.Start,
-		Last:      f.Last,
-	}
-
-	key := kernFlowKey(kernFlow)
-
-	f.UpdateUUID(key, flow.Opts{LayerKeyMode: flow.L3PreferedKeyMode})
-
-	fops = append(fops, &flow.Operation{
-		Type: flow.ReplaceOperation,
-		Flow: f,
-		Key:  key,
-	})
-	return fops
-}
-
-func (p *EBPFProbe) updateFlowOperation(ebpfFlow *EBPFFlow, kernFlow *C.struct_flow, startKTimeNs int64, start time.Time) *flow.Operation {
-	f := flow.NewFlow()
-	f.Last = common.UnixMillis(ebpfFlow.last)
-	layersInfo := uint8(kernFlow.layers_info)
-	if layersInfo&uint8(C.TRANSPORT_LAYER_INFO) > 0 {
-		protocol := uint8(kernFlow.transport_layer.protocol)
-		switch protocol {
-		case syscall.IPPROTO_TCP:
-			f.TCPMetric = &flow.TCPMetric{
-				ABSynStart: tcpFlagTime(kernFlow.transport_layer.ab_syn, startKTimeNs, start),
-				BASynStart: tcpFlagTime(kernFlow.transport_layer.ba_syn, startKTimeNs, start),
-				ABFinStart: tcpFlagTime(kernFlow.transport_layer.ab_fin, startKTimeNs, start),
-				BAFinStart: tcpFlagTime(kernFlow.transport_layer.ba_fin, startKTimeNs, start),
-				ABRstStart: tcpFlagTime(kernFlow.transport_layer.ab_rst, startKTimeNs, start),
-				BARstStart: tcpFlagTime(kernFlow.transport_layer.ba_rst, startKTimeNs, start),
-			}
-		}
-	}
-	f.Metric = &flow.FlowMetric{
-		ABBytes:   int64(kernFlow.metrics.ab_bytes),
-		ABPackets: int64(kernFlow.metrics.ab_packets),
-		BABytes:   int64(kernFlow.metrics.ba_bytes),
-		BAPackets: int64(kernFlow.metrics.ba_packets),
-		Start:     f.Start,
-		Last:      f.Last,
-	}
-
-	key := kernFlowKey(kernFlow)
-
-	return &flow.Operation{
-		Type: flow.UpdateOperation,
-		Flow: f,
-		Key:  key,
-	}
-}
-
 func (p *EBPFProbe) run() {
 	var info syscall.Sysinfo_t
 	syscall.Sysinfo(&info)
 
-	_, flowChanOperation := p.flowTable.Start()
+	_, flowChanOperation, _ := p.flowTable.Start()
 	defer p.flowTable.Stop()
 
 	var startKTimeNs int64
 	var start time.Time
 
-	ebpfFlows := make(map[C.__u64]*EBPFFlow)
 	updateTicker := time.NewTicker(ebpfUpdate)
 	defer updateTicker.Stop()
 
+	p.flow_pool_size = 2*cap(flowChanOperation) + 1
```

Comment:
why 2*cap ? the chan will block at cap so pool = cap+1 is enough

Reply:
this one is a bit tricky. It seems that go keeps local read/write buffers to use with channels so that when the channel blocks after cap entries and the reader performs a read operation the entire content of the channel is copied to its local buffer (for performance) and then reusing the pointers by the writer results in an overlap (I myself experienced this in one of the stress tests). therefore the pool size should be twice the size of the channel ( I agree that +1 here is redundant, can remove it if you'd like)

Reply:
golang channel use lockfree mechanism, but this should not be the double ... or I missing something ... do you have a bench test reproducing your issue (in case of cap+1) ?

Reply:
I found your issue and 2*cap()+1 will not fix it

Imaging you can push 3 elements in the chan
len(chan) is 3
next is 3
 ... now the chan reader side read the 3 elements
len(chan) is 0
next is still 3

so you are out of sync between the chan pointer and pointer that could be already referenced in another element in the chan.

so you should remove next and have another channel to send back, the freed pointers; at start the allocated pointers should be injected in the _freed_ channel.

i := getFreePointer()   // <- freeEbpfFlows
write in &ebpfFlows[i]
flowChanOperation <- &ebpfFlows[i]

Reply:
@nplanel @andrekassis is using circular buffer that the size of the buffer is twice the size of the channel he is writing to . You @nplanel are correct that next = 3 (in your example) but this is ok ... 3 is free to use ... and so is 4  and 5 (3*2 size) ... when new flows are written to 3,4,5 the reader will continue to read and this is ok because those are new flows ... when reader ends reading 3,4,5 next goes to 6 ... so circular goes to zero --- i.e. next = 0 and this is again ok, because we know that the reader already read 0,1 and 2 ... I think that 2 * capacity is ok (assuming the implementation of chan in golan has local buffer up to the size of the channel).

What you suggest will work ... I am sitting here with @andrekassis trying to review and I think this is fine. If we are missing something can you help by adding more elaborated example ... sorry for the long remark ... but we are trying to think and understand what is going on here.